### PR TITLE
NIAD-237: Mongodb persistence adapter

### PR DESCRIPTION
**Pull request by bartek-sarul**

### CODE REVIEW

File: common/persistence/dynamo_persistence_adaptor.py [hunk 1]
```diff
@@ -1,49 +1,19 @@
 """Module containing functionality for a DynamoDB implementation of a persistence adaptor."""
```
Comment (petervdm) [inline common/persistence/dynamo_persistence_adaptor.py]: These should be done in the new common repo

File: common/persistence/mongo_persistence_adaptor.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
+"""Module containing functionality for a DynamoDB implementation of a persistence adaptor."""
```
Comment (mattd-kainos) [inline common/persistence/mongo_persistence_adaptor.py]: should be mongo not dynamo

File: common/persistence/dynamo_persistence_adaptor.py (hunk 1 above)
Comment (petervdm) [reply common/persistence/dynamo_persistence_adaptor.py]: Do you want to hook up that common repo as part of this?

File: common/persistence/mongo_persistence_adaptor.py (hunk 2 above)
Comment (petervdm) [inline common/persistence/mongo_persistence_adaptor.py]: MongoDb

File: common/persistence/mongo_persistence_adaptor.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+
+class MongoPersistenceAdaptor(persistence_adaptor.PersistenceAdaptor):
+    """Class responsible for persisting items into a MongoDB."""
+
+    client = None
+
+    def __init__(self, table_name: str, max_retries: int, retry_delay: float):
+        """
+        Constructs a DynamoDB version of a
```
Comment (petervdm) [inline common/persistence/mongo_persistence_adaptor.py]: MongoDb

File: common/persistence/mongo_persistence_adaptor.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            return self._prepare_data_to_return(result)
+        except Exception as e:
+            raise RecordUpdateError from e
+
+    @retriable
+    async def get(self, key: str, **kwargs):
+        """
+        Retrieves an item from a specified table with a given key.
+        :param key: The key which identifies the item to get.
+        :param strongly_consistent_read: https://docs.amazonaws.cn/en_us/amazondynamodb/latest/developerguide/HowItWorks.ReadConsistency.html
```
Comment (petervdm) [inline common/persistence/mongo_persistence_adaptor.py]: Probably not relevant to this?

**Review by petervdm — DISMISSED** [review]

**Comment by mattd-kainos** [comment]
Haven't had a chance to get through it all yet. Would be good to update the docs on how to run and test using mongo instead of dynamo.

**Comment by mattd-kainos** [comment]
Also have we thought about ensuring both are tested. Should we run component tests for both? Want to make sure there is not regression in the adaptor not being used as often.

**Review by mattd-kainos — CHANGES_REQUESTED** [review]
Mostly questions about config and preventing regression. Haven't tried to run locally yet.

**Comment by bartek-sarul** [comment]
> Also have we thought about ensuring both are tested. Should we run component tests for both? Want to make sure there is not regression in the adaptor not being used as often.

I'll add tests that will run same set of tests on all adaptors

**Comment by bartek-sarul** [comment]
> Would be good to update the docs on how to run and test using mongo instead of dynamo.

I'll update config

File: common/persistence/dynamo_persistence_adaptor.py (hunk 1 above)
Comment (petervdm) [reply common/persistence/dynamo_persistence_adaptor.py]: Will do in a new ticket

File: common/persistence/dynamo_persistence_adaptor.py [hunk 5]
```diff
@@ -59,30 +29,33 @@ def __init__(self, table_name: str, max_retries: int, retry_delay: float):
           * retry_delay: The delay between retries
         :param table_name: Table name to be used in this adaptor.
         """
+        self.table_name = table_name
         self.retry_delay = retry_delay
         self.max_retries = max_retries
-        self.table_name = table_name
 
+        self.endpoint_url = config.get_config('DB_ENDPOINT_URL', None)
+        self.region_name = 'eu-west-2'
```
Comment (mattd-kainos) [inline common/persistence/dynamo_persistence_adaptor.py]: I don't think this should be hardcoded.

File: integration-tests/integration_tests/integration_tests/db/dynamo_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
+"""
+Provides a way of accessing a Dynamo table
+"""
+import os
+
+import boto3
+
+
+class DynamoWrapper:
```
Comment (mattd-kainos) [inline integration-tests/integration_tests/integration_tests/db/dynamo_wrapper.py]: Consider defining an interface for these wrappers

**Review by mattd-kainos — CHANGES_REQUESTED** [review]
Just one concern about the hardcoded AWS region that could cause headaches in the future. Optional improvement for wrapper interface - mostly for better autocomplete etc for devs in IDEs

File: common/persistence/mongo_persistence_adaptor.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    def _prepare_data_to_add(key, data):
+        item = copy.deepcopy(data)
+        item[_KEY] = key
+        return item
+
+    @staticmethod
+    def _prepare_data_to_return(data):
+        if data is not None:
+            del data[_KEY]
+        return data
```
Comment (grzegorz-robak) [inline common/persistence/mongo_persistence_adaptor.py]: These methods are duplicated in each persistence adaptor

File: common/persistence/persistence_adaptor.py [hunk 8]
```diff
@@ -4,14 +4,65 @@
...
 from typing import Optional
 
+from exceptions import MaxRetriesExceeded
+from retry.retriable_action import RetriableAction
+
+DATA_VALIDATION_ERROR_MESSAGE = "Data must not have field named '{}' as it's used " \
+                                 "as primary key and is explicitly set as this function argument"
+
+
+def validate_data(primary_key: str):
```
Comment (grzegorz-robak) [inline common/persistence/persistence_adaptor.py]: I'd name it like `data_without_field` or `data_without_key` as `validate_data` is ambiguous, when it does just one thing.

File: common/persistence/persistence_adaptor_factory.py [hunk 9]
```diff
@@ -1,13 +1,15 @@
...
+from persistence.mongo_persistence_adaptor import MongoPersistenceAdaptor
 from persistence.persistence_adaptor import PersistenceAdaptor
 
 logger = log.IntegrationAdaptorsLogger(__name__)
 
 _DEFAULT_ADAPTOR = 'dynamodb'
-_PERSISTENCE_ADAPTOR_TYPES = {
+PERSISTENCE_ADAPTOR_TYPES = {
     _DEFAULT_ADAPTOR: DynamoPersistenceAdaptor,
+    'mongodb': MongoPersistenceAdaptor,
```
Comment (grzegorz-robak) [inline common/persistence/persistence_adaptor_factory.py]: option to set DynamoDB explicitly should be added also

File: integration-tests/integration_tests/integration_tests/db/dynamo_wrapper.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        return DynamoWrapper(
+            table_name=os.environ.get('MHS_STATE_TABLE_NAME', 'mhs_state'),
+            region_name='eu-west-2',
+            endpoint_url=os.environ.get('MHS_DB_ENDPOINT_URL', None))
+
+    @staticmethod
+    def get_sync_async_table_wrapper():
+        return DynamoWrapper(
+            table_name=os.environ.get('MHS_SYNC_ASYNC_TABLE_NAME', 'sync_async_state'),
+            region_name='eu-west-2',
```
Comment (grzegorz-robak) [inline integration-tests/integration_tests/integration_tests/db/dynamo_wrapper.py]: hardcoded region

File: mhs/spineroutelookup/Pipfile [hunk 11]
```diff
@@ -9,11 +9,11 @@ unittest-xml-reporting = "~=2.5"
 
 [packages]
 integration-adaptors-common = {editable = true,path = "./../../common"}
-tornado = "*"
+tornado = "~=6.0"
 defusedxml = "~=0.6"
-ldap3 = "*"
+ldap3 = "~=2.6"
```
Comment (grzegorz-robak) [inline mhs/spineroutelookup/Pipfile]: that is different value than in integration tests

**Review by grzegorz-robak — CHANGES_REQUESTED** [review]

File: common/persistence/dynamo_persistence_adaptor.py [hunk 12]
```diff
@@ -59,30 +28,33 @@ def __init__(self, table_name: str, max_retries: int, retry_delay: float):
           * retry_delay: The delay between retries
         :param table_name: Table name to be used in this adaptor.
         """
+        self.table_name = table_name
         self.retry_delay = retry_delay
         self.max_retries = max_retries
-        self.table_name = table_name
 
+        self.endpoint_url = config.get_config('DB_ENDPOINT_URL', None)
+        self.region_name = config.get_config('MHS_CLOUD_REGION', 'eu-west-2')
```
Comment (mattd-kainos) [inline common/persistence/dynamo_persistence_adaptor.py]: Should just be 'CLOUD_REGION'. config will strip the MHS_ prefix

**Review by mattd-kainos — DISMISSED** [review]

**Review by grzegorz-robak — APPROVED** [review]

**Review by grzegorz-robak — DISMISSED** [review]

**Comment by bartek-sarul** [comment]
Force merge as PR has been approved before build fixes

**Pull request merged**